File: 606-construct-string-from-binary-tree/606-construct-string-from-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def tree2str(self, root: Optional[TreeNode]) -> str:
        if root == None: return ""
        st = str()
        
        st = str(root.val)
        if root.left or root.right:
            st = st + "(" + self.tree2str(root.left) + ")"
            if root.right:
                st = st + "(" + self.tree2str(root.right) + ")"
        
        return st
        

File: 606-construct-string-from-binary-tree/606-construct-string-from-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    string serialize(TreeNode* root) {
        string str = "";
        if(root == NULL) return str;
        if(root->right)     str = to_string(root->val) + "(" + serialize(root->left) + ")" + "(" + serialize(root->right) + ")";
        else if(root->left) str = to_string(root->val) + "(" + serialize(root->left) + ")";
        else                str = to_string(root->val);
        return str;
    }
public:
    string tree2str(TreeNode* root) {
        return serialize(root);
    }
};